src/geometry/utility/geometry.cc

```diff
@@ -293,123 +286,123 @@ CoordinateT InterpolatedPointAtP(const LineString<CoordinateT>& line_string, dou
-  const Vector3 d_xyz{*line_string_points_length.second - *line_string_points_length.first};
+  const auto bound_points = GetBoundPointsAtP(line_string, p, tolerance);
+  const Vector3 d_xyz{line_string[bound_points.idx_end] - line_string[bound_points.idx_start]};
   return (d_xyz / (line_string.length())).normalized();
 }
 
 template <typename CoordinateT>
 ClosestPointResult<CoordinateT> GetClosestPointToSegment(const std::pair<CoordinateT, CoordinateT>& segment,
                                                          const CoordinateT& coordinate, double tolerance) {
   const CoordinateT d_segment{segment.second - segment.first};
+  const CoordinateT d_segment_normalized{d_segment.normalized()};
   const CoordinateT d_coordinate_to_first{coordinate - segment.first};
 
-  const double unsaturated_p = d_coordinate_to_first.dot(d_segment.normalized());
+  const double unsaturated_p = d_coordinate_to_first.dot(d_segment_normalized);
   const double p = std::clamp(unsaturated_p, 0., d_segment.norm());
-  const CoordinateT point = InterpolatedPointAtP(LineString<CoordinateT>{segment.first, segment.second}, p, tolerance);
+
+  // point at p
+  const CoordinateT point = p * d_segment_normalized + segment.first;
   const double distance = (coordinate - point).norm();
   return {p, point, distance};
 }
 
 ClosestPointResult3d GetClosestPoint(const LineString3d& line_string, const maliput::math::Vector3& xyz,
                                      double tolerance) {
-  ClosestPointResult3d result;
-  result.distance = std::numeric_limits<double>::max();
-  double length{};
-  for (auto first = line_string.begin(), second = std::next(line_string.begin()); second != line_string.end();
-       ++first, ++second) {
-    // If points are under numeric tolerance, skip segment.
-    if ((*first - *second).norm() < kEpsilon) continue;
-    const auto closest_point_res = GetClosestPointToSegment(Segment3d{*first, *second}, xyz, tolerance);
-    if (closest_point_res.distance < result.distance) {
-      result = closest_point_res;
-      result.p += length;
+  std::optional<LineString3d::Segment> closest_segment{std::nullopt};
+  ClosestPointResult3d segment_closest_point_result;
+  segment_closest_point_result.distance = std::numeric_limits<double>::max();
+
+  const auto& segments = line_string.segments();
+  std::for_each(std::execution::par, segments.begin(), segments.end(), [&](const auto& segment) {
+    const auto& start = line_string[segment.second.idx_start];
+    const auto& end = line_string[segment.second.idx_end];
+    const auto current_closest_point_res = GetClosestPointToSegment(Segment3d{start, end}, xyz, tolerance);
+    if (current_closest_point_res.distance < segment_closest_point_result.distance) {
+      segment_closest_point_result = current_closest_point_res;
+      closest_segment = {segment.second.idx_start, segment.second.idx_end, segment.second.p_interval};
     }
```

Comment:
Sorry, this is not aligned with parallel execution as there might be race conditions, so I will remove the parallel execution here and use a regular for_each statement.

Reply:
Done at 9f81200